<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Classify duplicate gene pairs based on their modes of duplication"><title>Classify duplicate gene pairs based on their modes of duplication — classify_gene_pairs • doubletrouble</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.8/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Classify duplicate gene pairs based on their modes of duplication — classify_gene_pairs"><meta property="og:description" content="Classify duplicate gene pairs based on their modes of duplication"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">doubletrouble</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/doubletrouble_vignette.html">Identification and classification of duplicated genes</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/almeidasilvaf/doubletrouble/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Classify duplicate gene pairs based on their modes of duplication</h1>
      <small class="dont-index">Source: <a href="https://github.com/almeidasilvaf/doubletrouble/blob/HEAD/R/duplicate_classification.R" class="external-link"><code>R/duplicate_classification.R</code></a></small>
      <div class="d-none name"><code>classify_gene_pairs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Classify duplicate gene pairs based on their modes of duplication</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">classify_gene_pairs</span><span class="op">(</span></span>
<span>  annotation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  blast_list <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scheme <span class="op">=</span> <span class="st">"standard"</span>,</span>
<span>  blast_inter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">intron_counts</span>,</span>
<span>  evalue <span class="op">=</span> <span class="fl">1e-10</span>,</span>
<span>  anchors <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  max_gaps <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  proximal_max <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  collinearity_dir <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>annotation</dt>
<dd><p>A processed GRangesList or CompressedGRangesList object as
returned by <code><a href="https://rdrr.io/pkg/syntenet/man/process_input.html" class="external-link">syntenet::process_input()</a></code>.</p></dd>


<dt>blast_list</dt>
<dd><p>A list of data frames containing BLAST tabular output
for intraspecies comparisons.
Each list element corresponds to the BLAST output for a given species,
and names of list elements must match the names of list elements in
<strong>annotation</strong>. BLASTp, DIAMOND or simular programs must be run
on processed sequence data as returned by <code>process_input()</code>.</p></dd>


<dt>scheme</dt>
<dd><p>Character indicating which classification scheme to use.
One of "binary", "standard", "extended", or "full". See details below
for information on what each scheme means. Default: "standard".</p></dd>


<dt>blast_inter</dt>
<dd><p>(Only valid if <code>scheme == "extended" or "full"</code>).
A list of data frames containing BLAST tabular output
for the comparison between target species and outgroups.
Names of list elements must match the names of
list elements in <code>annotation</code>. BLASTp, DIAMOND or simular programs must
be run on processed sequence data as returned by <code>process_input()</code>.</p></dd>


<dt>intron_counts</dt>
<dd><p>(Only valid if <code>scheme == "full"</code>).
A list of 2-column data frames with the number of
introns per gene as returned by <code><a href="get_intron_counts.html">get_intron_counts()</a></code>. Names
of list elements must match names of <strong>annotation</strong>.</p></dd>


<dt>evalue</dt>
<dd><p>Numeric scalar indicating the E-value threshold.
Default: 1e-10.</p></dd>


<dt>anchors</dt>
<dd><p>Numeric indicating the minimum required number of genes
to call a syntenic block, as in <code><a href="https://rdrr.io/pkg/syntenet/man/infer_syntenet.html" class="external-link">syntenet::infer_syntenet</a></code>.
Default: 5.</p></dd>


<dt>max_gaps</dt>
<dd><p>Numeric indicating the number of upstream and downstream
genes to search for anchors, as in <code><a href="https://rdrr.io/pkg/syntenet/man/infer_syntenet.html" class="external-link">syntenet::infer_syntenet</a></code>.
Default: 25.</p></dd>


<dt>proximal_max</dt>
<dd><p>Numeric scalar with the maximum distance (in number
of genes) between two genes to consider them as proximal duplicates.
Default: 10.</p></dd>


<dt>collinearity_dir</dt>
<dd><p>Character indicating the path to the directory
where .collinearity files will be stored. If NULL, files will
be stored in a subdirectory of <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir()</a></code>. Default: NULL.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list of 3-column data frames of duplicated gene pairs
(columns 1 and 2), and their modes of duplication (column 3).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The classification schemes increase in complexity (number of classes)
in the order 'binary', 'standard', 'extended', and 'full'.</p>
<p>For classification scheme "binary", duplicates are classified into
one of 'SD' (segmental duplications) or 'SSD' (small-scale duplications).</p>
<p>For classification scheme "standard" (default), duplicates are
classified into 'SD' (segmental duplication), 'TD' (tandem duplication),
'PD' (proximal duplication), and 'DD' (dispersed duplication).</p>
<p>For classification scheme "extended", duplicates are classified into
'SD' (segmental duplication), 'TD' (tandem duplication),
'PD' (proximal duplication), 'TRD' (transposon-derived duplication),
and 'DD' (dispersed duplication).</p>
<p>Finally, for classification scheme "full", duplicates are classified into
'SD' (segmental duplication), 'TD' (tandem duplication),
'PD' (proximal duplication), 'rTRD' (retrotransposon-derived duplication),
'dTRD' (DNA transposon-derived duplication), and
'DD' (dispersed duplication).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load example data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">diamond_intra</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">diamond_inter</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">yeast_annot</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">yeast_seq</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get processed annotation data</span></span></span>
<span class="r-in"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu">syntenet</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/syntenet/man/process_input.html" class="external-link">process_input</a></span><span class="op">(</span><span class="va">yeast_seq</span>, <span class="va">yeast_annot</span><span class="op">)</span><span class="op">$</span><span class="va">annotation</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get list of intron counts</span></span></span>
<span class="r-in"><span><span class="va">txdb_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">yeast_annot</span>, <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGRanges.html" class="external-link">makeTxDbFromGRanges</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">intron_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">txdb_list</span>, <span class="va">get_intron_counts</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Classify duplicates - full scheme</span></span></span>
<span class="r-in"><span><span class="va">dup_class</span> <span class="op">&lt;-</span> <span class="fu">classify_gene_pairs</span><span class="op">(</span></span></span>
<span class="r-in"><span>    annotation <span class="op">=</span> <span class="va">annotation</span>, </span></span>
<span class="r-in"><span>    blast_list <span class="op">=</span> <span class="va">diamond_intra</span>, </span></span>
<span class="r-in"><span>    scheme <span class="op">=</span> <span class="st">"full"</span>,</span></span>
<span class="r-in"><span>    blast_inter <span class="op">=</span> <span class="va">diamond_inter</span>, </span></span>
<span class="r-in"><span>    intron_counts <span class="op">=</span> <span class="va">intron_counts</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Check number of gene pairs per class</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dup_class</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   SD   TD   PD rTRD dTRD   DD </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  342   42   80   52  963 2109 </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Fabrício Almeida-Silva, Yves Van de Peer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

