<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="doubletrouble">
<title>Identification and classification of duplicated genes • doubletrouble</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.2/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.2/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identification and classification of duplicated genes">
<meta property="og:description" content="doubletrouble">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#0092ac"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">doubletrouble</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/doubletrouble_vignette.html">Identification and classification of duplicated genes</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/almeidasilvaf/doubletrouble/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Identification and classification of duplicated genes</h1>
                        <h4 data-toc-skip class="author">Fabricio
Almeida-Silva</h4>
            <address class="author_afil">
      <p>VIB-UGent Center for Plant Systems Biology, Ghent University,
Ghent,
Belgium</p>
<br><h4 data-toc-skip class="author">Yves
Van de Peer</h4>
            <address class="author_afil">
      <p>VIB-UGent Center for Plant Systems Biology, Ghent University,
Ghent, Belgium</p>
<br><small class="dont-index">Source: <a href="https://github.com/almeidasilvaf/doubletrouble/blob/HEAD/vignettes/doubletrouble_vignette.Rmd" class="external-link"><code>vignettes/doubletrouble_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>doubletrouble_vignette.Rmd</code></div>
    </address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Gene and genome duplications are a source of raw genetic material for
evolution <span class="citation">(Ohno 2013)</span>. However,
whole-genome duplications (WGD) and small-scale duplications (SSD)
contribute to genome evolution in different manners. To help you explore
the different contributions of WGD and SSD to evolution, we developed
<em><a href="https://github.com/doubletrouble" class="external-link">doubletrouble</a></em>, a
package that can be used to identify and classify duplicated genes from
whole-genome protein sequences, calculate substitution rates per
substitution site (i.e., Ka and Ks) for gene pairs, find peaks in Ks
distributions, and classify gene pairs by age groups.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install <em><a href="https://github.com/doubletrouble" class="external-link">doubletrouble</a></em> from
Bioconductor with the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"doubletrouble"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Check that you have a valid Bioconductor installation</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/valid.html" class="external-link">valid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Then, you can load the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/almeidasilvaf/doubletrouble" class="external-link">doubletrouble</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-description">Data description<a class="anchor" aria-label="anchor" href="#data-description"></a>
</h2>
<p>In this vignette, we will use protein sequences (primary transcripts
only) and genome annotation for the yeast species <em>Saccharomyces
cerevisiae</em> and <em>Candida glabrata</em>. Data were obtained from
Ensembl Fungi release 54 <span class="citation">(Yates et al.
2022)</span>.</p>
<p>The example data sets are stored in the following objects:</p>
<ul>
<li>
<strong>yeast_seq:</strong> A list of <code>AAStringSet</code>
objects with elements named <em>Scerevisiae</em> and
<em>Cglabrata</em>.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load list of DIAMOND tabular output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">yeast_seq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">yeast_seq</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<strong>yeast_annot:</strong> A <code>GRangesList</code> object with
elements named <em>Scerevisiae</em> and <em>Cglabrata</em>.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load annotation list processed with syntenet::process_input()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">yeast_annot</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">yeast_annot</span><span class="op">)</span></span></code></pre></div>
<p>Our goal here is to identify and classify duplicated genes in the
<em>S. cerevisiae</em> genome. The <em>C. glabrata</em> genome will be
used as an outgroup to identify transposed duplicates later in this
vignette.</p>
</div>
<div class="section level2">
<h2 id="data-preparation">Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h2>
<p>First of all, we need to process the list of protein sequences and
gene ranges to detect synteny with <em><a href="https://bioconductor.org/packages/3.16/syntenet" class="external-link">syntenet</a></em>.
We will do that using the function <code><a href="https://rdrr.io/pkg/syntenet/man/process_input.html" class="external-link">process_input()</a></code> from the
<em><a href="https://bioconductor.org/packages/3.16/syntenet" class="external-link">syntenet</a></em>
package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/almeidasilvaf/syntenet" class="external-link">syntenet</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Process input data</span></span>
<span><span class="va">pdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/syntenet/man/process_input.html" class="external-link">process_input</a></span><span class="op">(</span><span class="va">yeast_seq</span>, <span class="va">yeast_annot</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pdata</span><span class="op">)</span></span>
<span><span class="va">pdata</span><span class="op">$</span><span class="va">seq</span></span>
<span><span class="va">pdata</span><span class="op">$</span><span class="va">annotation</span></span></code></pre></div>
<p>The processed data is represented as a list with the elements
<code>seq</code> and <code>annotation</code>, each containing the
protein sequences and gene ranges for each species, respectively.</p>
<p>Finally, we need to perform pairwise sequence similarity searches to
identify the whole set of paralogous gene pairs. We can do this using
the function <code><a href="https://rdrr.io/pkg/syntenet/man/run_diamond.html" class="external-link">run_diamond()</a></code> from the <em><a href="https://bioconductor.org/packages/3.16/syntenet" class="external-link">syntenet</a></em>
package <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;strong&gt;Note:&lt;/strong&gt; you need to have DIAMOND
installed in your machine to run this function. If you don’t have it,
you can use the &lt;em&gt;&lt;a href="https://bioconductor.org/packages/3.16/Herper" class="external-link"&gt;Herper&lt;/a&gt;&lt;/em&gt;
package to install DIAMOND in a Conda environment and run DIAMOND from
this virtual environment.&lt;/p&gt;'><sup>1</sup></a>, setting
<code>compare = "intraspecies"</code> to perform only intraspecies
comparisons.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">diamond_intra</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run DIAMOND in sensitive mode for S. cerevisiae only</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/syntenet/man/diamond_is_installed.html" class="external-link">diamond_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">diamond_intra</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/syntenet/man/run_diamond.html" class="external-link">run_diamond</a></span><span class="op">(</span></span>
<span>        seq <span class="op">=</span> <span class="va">pdata</span><span class="op">$</span><span class="va">seq</span><span class="op">[</span><span class="st">"Scerevisiae"</span><span class="op">]</span>,</span>
<span>        compare <span class="op">=</span> <span class="st">"intraspecies"</span>, </span>
<span>        outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"diamond_intra"</span><span class="op">)</span>,</span>
<span>        ... <span class="op">=</span> <span class="st">"--sensitive"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Inspect output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">diamond_intra</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">diamond_intra</span><span class="op">$</span><span class="va">Scerevisiae_Scerevisiae</span><span class="op">)</span></span></code></pre></div>
<p>And voilà! Now that we have the DIAMOND output and the processed
annotation, you can classify the duplicated genes.</p>
</div>
<div class="section level2">
<h2 id="classifying-duplicated-gene-pairs-and-genes">Classifying duplicated gene pairs and genes<a class="anchor" aria-label="anchor" href="#classifying-duplicated-gene-pairs-and-genes"></a>
</h2>
<p>To classify duplicated gene pairs based on their modes of
duplication, you will use the function
<code><a href="../reference/classify_gene_pairs.html">classify_gene_pairs()</a></code>. As input, you need to pass:</p>
<ul>
<li>The processed annotation list as returned by
<code><a href="https://rdrr.io/pkg/syntenet/man/process_input.html" class="external-link">syntenet::process_input()</a></code>;</li>
<li>A list of data frames with DIAMOND (or BLASTp, etc.) tabular output
as returned by <code><a href="https://rdrr.io/pkg/syntenet/man/run_diamond.html" class="external-link">syntenet::run_diamond()</a></code>.</li>
</ul>
<p>The function <code>classify_gene_pairs</code> incorporates 3
classification schemes:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Binary:</strong> Duplicated gene pairs are classified as
either derived from whole-genome duplications (WGD) or small-scale
duplications (SSD).</li>
<li>
<strong>Expanded:</strong> SSD-derived genes are divided into 3
subcategories: tandem duplication (TD), proximal duplication (PD), and
dispersed duplication (DD).</li>
<li>
<strong>Full:</strong> Similar to the <strong>expanded</strong>
scheme, but an additional subcategory is used for SSD-derived gene
pairs: transposed duplication (TRD). <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;&lt;strong&gt;Note:&lt;/strong&gt; To identify duplicated genes
derived from transpositions, you will need to input the DIAMOND output
of a similarity search between your species of interest and an outgroup.
See Section @ref(full) for details.&lt;/p&gt;"><sup>2</sup></a>
</li>
</ol>
<div class="section level3">
<h3 id="binary-classification">Binary classification<a class="anchor" aria-label="anchor" href="#binary-classification"></a>
</h3>
<p>To classify duplicated gene pairs into WGD- or SSD-derived, use
<code>binary = TRUE</code> in <code><a href="../reference/classify_gene_pairs.html">classify_gene_pairs()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Binary classification</span></span>
<span><span class="va">c_binary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify_gene_pairs.html">classify_gene_pairs</a></span><span class="op">(</span></span>
<span>    blast_list <span class="op">=</span> <span class="va">diamond_intra</span>,</span>
<span>    annotation <span class="op">=</span> <span class="va">pdata</span><span class="op">$</span><span class="va">annotation</span>,</span>
<span>    binary <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspecting the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">c_binary</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">c_binary</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How many pairs are there for each duplication mode?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">c_binary</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span></code></pre></div>
<p>The function returns a list of data frames, each containing the
duplicated gene pairs and their modes of duplication for each
species.</p>
</div>
<div class="section level3">
<h3 id="expanded-classification-ssd-td-pd-dd">Expanded classification (SSD → TD, PD, DD)<a class="anchor" aria-label="anchor" href="#expanded-classification-ssd-td-pd-dd"></a>
</h3>
<p>To subdivide SSD-derived gene pairs into TD-, PD-, and DD-derived,
use <code>binary = FALSE</code> in
<code><a href="../reference/classify_gene_pairs.html">classify_gene_pairs()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Expanded classification</span></span>
<span><span class="va">c_expanded</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify_gene_pairs.html">classify_gene_pairs</a></span><span class="op">(</span></span>
<span>    blast_list <span class="op">=</span> <span class="va">diamond_intra</span>,</span>
<span>    annotation <span class="op">=</span> <span class="va">pdata</span><span class="op">$</span><span class="va">annotation</span>,</span>
<span>    binary <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspecting the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">c_expanded</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">c_expanded</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How many pairs are there for each duplication mode?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">c_expanded</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="full">Full classification (SSD → TD, PD, TRD, DD)<a class="anchor" aria-label="anchor" href="#full"></a>
</h3>
<p>To find transposon-derived duplicates, first you will need to perform
a similarity search of your target species against an outgroup species.
You can do this with <code><a href="https://rdrr.io/pkg/syntenet/man/run_diamond.html" class="external-link">syntenet::run_diamond()</a></code>. For the
parameter <code>compare</code>, you will pass a 2-column data frame
specifying the comparisons to be made. <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;&lt;strong&gt;Pro tip:&lt;/strong&gt; If you want to identify and
classify duplicated genes for multiple species in batch, you must
include the outgroup for each of them in the comparisons data frame.&lt;/p&gt;"><sup>3</sup></a></p>
<p>Here, we will identify duplicated gene pairs for <em>Saccharomyces
cerevisiae</em> using <em>Candida glabrata</em> as an outgroup.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">diamond_inter</span><span class="op">)</span> <span class="co"># load pre-computed output in case DIAMOND is not installed</span></span>
<span></span>
<span><span class="co"># Create data frame of comparisons to be made</span></span>
<span><span class="va">comparisons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    species <span class="op">=</span> <span class="st">"Scerevisiae"</span>,</span>
<span>    outgroup <span class="op">=</span> <span class="st">"Cglabrata"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">comparisons</span></span>
<span></span>
<span><span class="co"># Run DIAMOND for the comparison we specified</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/syntenet/man/diamond_is_installed.html" class="external-link">diamond_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">diamond_inter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/syntenet/man/run_diamond.html" class="external-link">run_diamond</a></span><span class="op">(</span></span>
<span>        seq <span class="op">=</span> <span class="va">pdata</span><span class="op">$</span><span class="va">seq</span>,</span>
<span>        compare <span class="op">=</span> <span class="va">comparisons</span>,</span>
<span>        outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"diamond_inter"</span><span class="op">)</span>,</span>
<span>        ... <span class="op">=</span> <span class="st">"--sensitive"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">diamond_inter</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">diamond_inter</span><span class="op">$</span><span class="va">Scerevisiae_Cglabrata</span><span class="op">)</span></span></code></pre></div>
<p>Now, we will pass this interspecies DIAMOND output as an argument to
the parameter <code>blast_inter</code> of
<code><a href="../reference/classify_gene_pairs.html">classify_gene_pairs()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Full classification</span></span>
<span><span class="va">c_full</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify_gene_pairs.html">classify_gene_pairs</a></span><span class="op">(</span></span>
<span>    blast_list <span class="op">=</span> <span class="va">diamond_intra</span>,</span>
<span>    annotation <span class="op">=</span> <span class="va">pdata</span><span class="op">$</span><span class="va">annotation</span>,</span>
<span>    binary <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    blast_inter <span class="op">=</span> <span class="va">diamond_inter</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspecting the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">c_full</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">c_full</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How many pairs are there for each duplication mode?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">c_full</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="classifying-genes-into-unique-modes-of-duplication">Classifying genes into unique modes of duplication<a class="anchor" aria-label="anchor" href="#classifying-genes-into-unique-modes-of-duplication"></a>
</h2>
<p>If you look carefully at the output of
<code><a href="../reference/classify_gene_pairs.html">classify_gene_pairs()</a></code>, you will notice that some genes
appear in more than one duplicate pair, and these pairs can have
different duplication modes assigned. There’s nothing wrong with it.
Consider, for example, a gene that was originated by a whole-genome
duplication event some 60 million years ago, and then it underwent a
tandem duplication 5 million years ago. In the output of
<code><a href="../reference/classify_gene_pairs.html">classify_gene_pairs()</a></code>, you’d see this gene in two pairs,
one with <strong>WGD</strong> in the <em>type</em> column, and one with
<strong>TD</strong>.</p>
<p>If you want to assign each gene to a unique mode of duplication, you
can use the function <code><a href="../reference/classify_genes.html">classify_genes()</a></code>. This function
assigns duplication modes hierarchically in this level of priority: WGD
&gt; TD &gt; PD &gt; TRD &gt; DD.</p>
<p>The input for <code><a href="../reference/classify_genes.html">classify_genes()</a></code> is the list of gene pairs
returned by <code><a href="../reference/classify_gene_pairs.html">classify_gene_pairs()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Classify genes into unique modes of duplication</span></span>
<span><span class="va">c_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify_genes.html">classify_genes</a></span><span class="op">(</span><span class="va">c_full</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspecting the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">c_genes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">c_genes</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Number of genes per mode</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">c_genes</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculating-ka-ks-and-kaks-for-duplicated-gene-pairs">Calculating Ka, Ks, and Ka/Ks for duplicated gene pairs<a class="anchor" aria-label="anchor" href="#calculating-ka-ks-and-kaks-for-duplicated-gene-pairs"></a>
</h2>
<p>You can use the function <code><a href="../reference/pairs2kaks.html">pairs2kaks()</a></code> to calculate rates
of nonsynonymous substitutions per nonsynonymous site (Ka), synonymouys
substitutions per synonymous site (Ks), and their ratios (Ka/Ks). These
rates are calculated using the Bioconductor package <em><a href="https://bioconductor.org/packages/3.16/MSA2dist" class="external-link">MSA2dist</a></em>,
which implements all codon models in KaKs_Calculator 2.0 <span class="citation">(Wang et al. 2010)</span>.</p>
<p>For the purpose of demonstration, we will only calculate Ka, Ks, and
Ka/Ks for 5 WGD-derived gene pairs. The CDS for WGD-derived genes were
obtained from Ensembl Fungi <span class="citation">(Yates et al.
2022)</span>, and they are stored in <code>cds_scerevisiae</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cds_scerevisiae</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cds_scerevisiae</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Store DNAStringSet object in a list</span></span>
<span><span class="va">cds_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Scerevisiae <span class="op">=</span> <span class="va">cds_scerevisiae</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Keep only top give gene pairs for demonstration purposes</span></span>
<span><span class="va">gene_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Scerevisiae <span class="op">=</span> <span class="va">c_binary</span><span class="op">$</span><span class="va">Scerevisiae</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate Ka, Ks, and Ka/Ks</span></span>
<span><span class="va">kaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pairs2kaks.html">pairs2kaks</a></span><span class="op">(</span><span class="va">gene_pairs</span>, <span class="va">cds_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">kaks</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identifying-and-visualizing-ks-peaks">Identifying and visualizing Ks peaks<a class="anchor" aria-label="anchor" href="#identifying-and-visualizing-ks-peaks"></a>
</h2>
<p>Peaks in Ks distributions typically indicate whole-genome duplication
(WGD) events, and they can be identified by fitting Gaussian mixture
models (GMMs) to Ks distributions. In <em><a href="https://github.com/doubletrouble" class="external-link">doubletrouble</a></em>, this can
be performed with the function <code><a href="../reference/find_ks_peaks.html">find_ks_peaks()</a></code>.</p>
<p>However, because of saturation at higher Ks values, only
<strong>recent WGD</strong> events can be reliably identified from Ks
distributions <span class="citation">(Vanneste, Van de Peer, and Maere
2013)</span>. Recent WGD events are commonly found in plant species,
such as maize, soybean, apple, etc. Although the genomes of yeast
species have signatures of WGD, these events are ancient, so it is very
hard to find evidence for them using Ks distributions. <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;strong&gt;Tip:&lt;/strong&gt; You might be asking yourself: “How
does one identify ancient WGD, then?”. A common approach is to look for
syntenic blocks (i.e., regions with conserved gene content and order)
within genomes. This is what &lt;code&gt;&lt;a href="../reference/classify_gene_pairs.html"&gt;classify_gene_pairs()&lt;/a&gt;&lt;/code&gt; does
under the hood to find WGD-derived gene pairs.&lt;/p&gt;'><sup>4</sup></a></p>
<p>To demonstrate how you can find peaks in Ks distributions with
<code><a href="../reference/find_ks_peaks.html">find_ks_peaks()</a></code>, we will use a data frame containing Ks
values for duplicate pairs in the soybean (<em>Glycine max</em>) genome,
which has undergone 2 WGDs events ~13 and ~58 million years ago <span class="citation">(Schmutz et al. 2010)</span>. Then, we will visualize
Ks distributions with peaks using the function
<code><a href="../reference/plot_ks_peaks.html">plot_ks_peaks()</a></code>.</p>
<p>First of all, let’s look at the data and have a quick look at the
distribution.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load data and inspect it</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gmax_ks</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gmax_ks</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot distribution</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gmax_ks</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Ks</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="st">"grey90"</span>, bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>By visual inspection, we can see 2 or 3 peaks. Based on our prior
knowledge, we know that 2 WGD events have occurred in the ancestral of
the <em>Glycine</em> genus and in the ancestral of all Fabaceae, which
seem to correspond to the peaks we see at Ks values around 0.1 and 0.5,
respectively. There could be a third, flattened peak at around 1.6,
which would represent the WGD shared by all eudicots. Let’s test which
number of peaks has more support: 2 or 3.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find 2 and 3 peaks and test which one has more support</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_ks_peaks.html">find_ks_peaks</a></span><span class="op">(</span><span class="va">gmax_ks</span><span class="op">$</span><span class="va">Ks</span>, npeaks <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize Ks distribution</span></span>
<span><span class="fu"><a href="../reference/plot_ks_peaks.html">plot_ks_peaks</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<p>As we can see, the presence of 3 peaks is more supported (lowest
BIC). The function returns a list with the mean, variance and amplitude
of mixture components (i.e., peaks), as well as the Ks distribution
itself.</p>
<p>Now, suppose you just want to get the first 2 peaks. You can do that
by explictly saying to <code><a href="../reference/find_ks_peaks.html">find_ks_peaks()</a></code> how many peaks there
are.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find 2 peaks ignoring Ks values &gt; 1</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_ks_peaks.html">find_ks_peaks</a></span><span class="op">(</span><span class="va">gmax_ks</span><span class="op">$</span><span class="va">Ks</span>, npeaks <span class="op">=</span> <span class="fl">2</span>, max_ks <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_ks_peaks.html">plot_ks_peaks</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<p><strong>Important consideration on GMMs and Ks
distributions:</strong> Peaks identified with GMMs should not be blindly
regarded as “the truth”. Using GMMs to find peaks in Ks distributions
can lead to problems such as overfitting and overclustering <span class="citation">(Tiley, Barker, and Burleigh 2018)</span>. Some general
recommendations are:</p>
<ol style="list-style-type: decimal">
<li><p>Use your prior knowledge. If you know how many peaks there are
(e.g., based on literature evidence), just tell the number to
<code><a href="../reference/find_ks_peaks.html">find_ks_peaks()</a></code>. Likewise, if you are not sure about how
many peaks there are, but you know the maximum number of peaks is N,
don’t test for the presence of &gt;N peaks. GMMs can incorrectly
identify more peaks than the actual number.</p></li>
<li><p>Test the significance of each peak with SiZer (Significant ZERo
crossings of derivatives) maps <span class="citation">(Chaudhuri and
Marron 1999)</span>. This can be done with the function
<code>SiZer()</code> from the R package <em><a href="https://CRAN.R-project.org/package=feature" class="external-link">feature</a></em>.</p></li>
</ol>
<p>As an example of a SiZer map, let’s use <code><a href="https://rdrr.io/pkg/feature/man/SiZer.html" class="external-link">feature::SiZer()</a></code>
to assess the significance of the 2 peaks we found previously.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get numeric vector of Ks values &lt;= 1</span></span>
<span><span class="va">ks</span> <span class="op">&lt;-</span> <span class="va">gmax_ks</span><span class="op">$</span><span class="va">Ks</span><span class="op">[</span><span class="va">gmax_ks</span><span class="op">$</span><span class="va">Ks</span> <span class="op">&lt;=</span> <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Get SiZer map</span></span>
<span><span class="fu">feature</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/feature/man/SiZer.html" class="external-link">SiZer</a></span><span class="op">(</span><span class="va">ks</span><span class="op">)</span></span></code></pre></div>
<p>The blue regions in the SiZer map indicate significantly increasing
regions of the curve, which support the 2 peaks we found.</p>
</div>
<div class="section level2">
<h2 id="classifying-genes-by-age-groups">Classifying genes by age groups<a class="anchor" aria-label="anchor" href="#classifying-genes-by-age-groups"></a>
</h2>
<p>Finally, you can use the peaks you obtained before to classify gene
pairs by age group. Age groups are defined based on the Ks peak to which
pairs belong. This is useful if you want to analyze duplicate pairs from
a specific WGD event, for instance. You can do this with the function
<code><a href="../reference/split_pairs_by_peak.html">split_pairs_by_peak()</a></code>. This function returns a list
containing the classified pairs in a data frame, and a ggplot object
with the age boundaries highlighted in the histogram of Ks values.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Gene pairs without age-based classification</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gmax_ks</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Classify gene pairs by age group</span></span>
<span><span class="va">pairs_age_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/split_pairs_by_peak.html">split_pairs_by_peak</a></span><span class="op">(</span><span class="va">gmax_ks</span>, <span class="va">peaks</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspecting the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pairs_age_group</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look at the classified gene pairs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pairs_age_group</span><span class="op">$</span><span class="va">pairs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize Ks distro with age boundaries</span></span>
<span><span class="va">pairs_age_group</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This document was created under the following conditions:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-chaudhuri1999sizer" class="csl-entry">
Chaudhuri, Probal, and James S Marron. 1999. <span>“SiZer for
Exploration of Structures in Curves.”</span> <em>Journal of the American
Statistical Association</em> 94 (447): 807–23.
</div>
<div id="ref-ohno2013evolution" class="csl-entry">
Ohno, Susumu. 2013. <em>Evolution by Gene Duplication</em>. Springer
Science &amp; Business Media.
</div>
<div id="ref-schmutz2010genome" class="csl-entry">
Schmutz, Jeremy, Steven B Cannon, Jessica Schlueter, Jianxin Ma, Therese
Mitros, William Nelson, David L Hyten, et al. 2010. <span>“Genome
Sequence of the Palaeopolyploid Soybean.”</span> <em>Nature</em> 463
(7278): 178–83.
</div>
<div id="ref-tiley2018assessing" class="csl-entry">
Tiley, George P, Michael S Barker, and J Gordon Burleigh. 2018.
<span>“Assessing the Performance of Ks Plots for Detecting Ancient Whole
Genome Duplications.”</span> <em>Genome Biology and Evolution</em> 10
(11): 2882–98.
</div>
<div id="ref-vanneste2013inference" class="csl-entry">
Vanneste, Kevin, Yves Van de Peer, and Steven Maere. 2013.
<span>“Inference of Genome Duplications from Age Distributions
Revisited.”</span> <em>Molecular Biology and Evolution</em> 30 (1):
177–90.
</div>
<div id="ref-wang2010kaks_calculator" class="csl-entry">
Wang, Dapeng, Yubin Zhang, Zhang Zhang, Jiang Zhu, and Jun Yu. 2010.
<span>“KaKs_calculator 2.0: A Toolkit Incorporating Gamma-Series Methods
and Sliding Window Strategies.”</span> <em>Genomics, Proteomics &amp;
Bioinformatics</em> 8 (1): 77–80.
</div>
<div id="ref-yates2022ensembl" class="csl-entry">
Yates, Andrew D, James Allen, Ridwan M Amode, Andrey G Azov, Matthieu
Barba, Andrés Becerra, Jyothish Bhai, et al. 2022. <span>“Ensembl
Genomes 2022: An Expanding Genome Resource for Non-Vertebrates.”</span>
<em>Nucleic Acids Research</em> 50 (D1): D996–1003.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Fabrício Almeida-Silva, Yves Van de Peer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
